File: api/_plan_guard.py
"""Plan checks shared by Python serverless functions.

Frontend gating is useful for UX, but paid endpoints also need a server-side
check so direct API calls cannot bypass the plan.
"""

import json
import os
import time
import urllib.error
import urllib.parse
import urllib.request


DEFAULT_SUPABASE_URL = "https://dhirxfoxcswctxcjzvhf.supabase.co"
ACTIVE_STATUS = {"active", "trialing"}
PLAN_RANK = {"free": 0, "basic": 1, "pro": 2, "premium": 3}
FEATURE_MIN_PLAN = {
    "modal": "basic",
    "live": "pro",
    "injuries": "pro",
    "football": "pro",
    "sports": "pro",
    "props_by_game": "premium",
}
_CACHE = {}
_CACHE_TTL = 60


def check_feature(headers, feature):
    """Return (ok, status, payload) for the requested feature."""
    min_plan = FEATURE_MIN_PLAN.get(feature)
    if not min_plan:
        return True, 200, {}

    try:
        user = _load_user_and_subscription(headers)
    except PlanError as error:
        return False, error.status, {"error": error.message, "feature": feature}

    if user.get("role") == "admin":
        return True, 200, user

    if user.get("status") not in ACTIVE_STATUS:
        return False, 403, {"error": "subscription inactive", "feature": feature}

    plan = user.get("plan") or "free"
    if PLAN_RANK.get(plan, 0) < PLAN_RANK[min_plan]:
        return False, 403, {
            "error": "plan upgrade required",
            "feature": feature,
            "required_plan": min_plan,
            "current_plan": plan,
        }

    return True, 200, user


def _load_user_and_subscription(headers):
    service_key = os.environ.get("SUPABASE_SERVICE_KEY", "").strip()
    if not service_key:
        raise PlanError("SUPABASE_SERVICE_KEY is not configured", 500)

    token = _extract_bearer(headers)
    if not token:
        raise PlanError("missing bearer token", 401)

    cache_key = f"plan:{token[-24:]}"
    cached = _CACHE.get(cache_key)
    if cached and time.time() < cached["exp"]:
        return cached["data"]

    supabase_url = _normalize_supabase_url(os.environ.get("SUPABASE_URL", ""))
    auth_user = _json_request(
        f"{supabase_url}/auth/v1/user",
        headers={
            "Accept": "application/json",
            "apikey": service_key,
            "Authorization": f"Bearer {token}",
        },
    )
    user_id = auth_user.get("id")
    if not user_id:
        raise PlanError("invalid token payload", 401)

    rows = _json_request(
        f"{supabase_url}/rest/v1/subscriptions?select=plan,status,role&user_id=eq.{urllib.parse.quote(user_id)}&limit=1",
        headers={
            "Accept": "application/json",
            "apikey": service_key,
            "Authorization": f"Bearer {service_key}",
        },
    )
    row = rows[0] if isinstance(rows, list) and rows else {}
    data = {
        "user_id": user_id,
        "email": auth_user.get("email"),
        "plan": row.get("plan") or "free",
        "status": row.get("status") or "active",
        "role": row.get("role") or "user",
    }
    _CACHE[cache_key] = {"data": data, "exp": time.time() + _CACHE_TTL}
    return data


def _json_request(url, headers):
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=8) as response:
            raw = response.read().decode("utf-8", errors="replace")
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as error:
        status = 401 if error.code in (401, 403) else 500
        raise PlanError("invalid bearer token" if status == 401 else f"Supabase HTTP {error.code}", status)


def _extract_bearer(headers):
    value = headers.get("Authorization") or headers.get("authorization") or ""
    value = str(value).strip()
    if value.lower().startswith("bearer "):
        return value[7:].strip()
    return value


def _normalize_supabase_url(value):
    raw = str(value or "").strip().rstrip("/")
    if raw.startswith("https://") and raw.endswith(".supabase.co"):
        return raw
    return DEFAULT_SUPABASE_URL


class PlanError(Exception):
    def __init__(self, message, status):
        super().__init__(message)
        self.message = message
        self.status = status

File: api/_planGuard.js
import { createHash } from 'node:crypto';
import {
  extractBearerToken,
  normalizeSupabaseUrl,
  serviceHeaders,
  supabaseFetch,
  verifySupabaseToken,
} from './_supabaseAuth.js';

const ACTIVE = new Set(['active', 'trialing']);
const RANK = { free: 0, basic: 1, pro: 2, premium: 3 };
const REQUIRED = { modal: 'basic', live: 'pro', injuries: 'pro', football: 'pro', sports: 'pro', cs2: 'pro', props_by_game: 'premium' };
const accessCache = new Map();
const inflightAccess = new Map();
const ACCESS_TTL_MS = 30_000;
const MAX_ACCESS_ENTRIES = 500;

export async function checkFeature(req, feature) {
  const requiredPlan = REQUIRED[feature];
  if (!requiredPlan) return { ok: true };
  try {
    const access = await loadUserAccess(req.headers.authorization);
    if (access.role === 'admin') return { ok: true, access };
    if (!ACTIVE.has(access.status)) return { ok: false, status: 403, payload: { error: 'subscription inactive', feature } };
    if ((RANK[access.plan] || 0) < RANK[requiredPlan]) {
      return { ok: false, status: 403, payload: { error: 'plan upgrade required', feature, required_plan: requiredPlan, current_plan: access.plan } };
    }
    return { ok: true, access };
  } catch (error) {
    const status = [401, 403, 500, 503].includes(error.status) ? error.status : 503;
    const message = status === 401 ? error.message : status === 500 ? 'authorization service is not configured' : 'authorization service unavailable';
    return { ok: false, status, payload: { error: message, feature } };
  }
}

export async function loadUserAccess(authorization, config = {}) {
  const token = extractBearerToken(authorization);
  if (!token) throw httpError('missing bearer token', 401);
  const supabaseUrl = normalizeSupabaseUrl(config.supabaseUrl ?? process.env.SUPABASE_URL);
  const serviceKey = String(config.serviceKey ?? process.env.SUPABASE_SERVICE_KEY ?? '').trim();
  if (!serviceKey) throw httpError('SUPABASE_SERVICE_KEY is not configured', 500);
  const cacheKey = createHash('sha256').update(token).digest('hex');
  const cached = accessCache.get(cacheKey);
  if (cached && Date.now() - cached.savedAt < ACCESS_TTL_MS) return cached.data;
  if (cached) accessCache.delete(cacheKey);
  if (inflightAccess.has(cacheKey)) return inflightAccess.get(cacheKey);

  const request = fetchUserAccess({ serviceKey, supabaseUrl, token }).then((data) => {
    if (inflightAccess.get(cacheKey) === request) {
      accessCache.set(cacheKey, { data, savedAt: Date.now() });
      trimAccessCache();
    }
    return data;
  }).finally(() => {
    if (inflightAccess.get(cacheKey) === request) inflightAccess.delete(cacheKey);
  });
  inflightAccess.set(cacheKey, request);
  return request;
}

export function clearPlanAccessCache() {
  accessCache.clear();
  inflightAccess.clear();
}

async function fetchUserAccess({ serviceKey, supabaseUrl, token }) {
  const user = await verifySupabaseToken(supabaseUrl, serviceKey, `Bearer ${token}`);
  const rows = await supabaseFetch(
    supabaseUrl,
    `/rest/v1/subscriptions?select=plan,status,role&user_id=eq.${encodeURIComponent(user.id)}&limit=1`,
    { headers: serviceHeaders(serviceKey) },
  );
  const subscription = Array.isArray(rows) && rows.length ? rows[0] : {};
  return {
    user_id: user.id,
    email: user.email,
    plan: subscription.plan || 'free',
    status: subscription.status || 'active',
    role: subscription.role || 'user',
  };
}

function trimAccessCache() {
  while (accessCache.size > MAX_ACCESS_ENTRIES) accessCache.delete(accessCache.keys().next().value);
}

function httpError(message, status) {
  return Object.assign(new Error(message), { status });
}
